### PR TITLE
Fixes #651 Adds Java modules support via module-info.java

### LINKED ISSUE #651: Support for the Java module system?
Hello everyone. My team is considering using RoaringBitmap in our product (a plugin for elasticsearch) but we are encountering some issue at runtime, which I believe is related to the java module system that elasticsearch adopts. I wonder if to modularize RoaringBitmap would help with this.

```
java.lang.module.ResolutionException: Modules shims and RoaringBitmap export package org.roaringbitmap to module java.uuid.generator
	at java.base/java.lang.module.Resolver.resolveFail(Resolver.java:900)
	at java.base/java.lang.module.Resolver.failTwoSuppliers(Resolver.java:814)
	at java.base/java.lang.module.Resolver.checkExportSuppliers(Resolver.java:735)
	at java.base/java.lang.module.Resolver.finish(Resolver.java:380)
	at java.base/java.lang.module.Configuration.<init>(Configuration.java:139)
	at java.base/java.lang.module.Configuration.resolveAndBind(Configuration.java:493)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.createModuleLayer(PluginsService.java:705)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.createPluginModuleLayer(PluginsService.java:673)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.createPlugin(PluginsService.java:527)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.loadBundle(PluginsService.java:451)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.loadBundles(PluginsService.java:292)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:161)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.plugins.PluginsService.lambda$getPluginsServiceCtor$15(PluginsService.java:668)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.node.Node.<init>(Node.java:417)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.node.Node.<init>(Node.java:324)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.bootstrap.Elasticsearch$2.<init>(Elasticsearch.java:216)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:216)
	at org.elasticsearch.server@8.7.1/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:67)
```

For what it matters, the issue my team encountered seems similar to the one reported in [this stackoverflow question](https://stackoverflow.com/questions/71288102/java-11-gradle-with-modules-javafx-tablesaw-yields-resolutionexception).

Thanks for your attention,
Matteo

### PR DESCRIPTION
**Pull request by stiemannkj1**
### SUMMARY

Adds `module-info.java` file to project to support deploying RoaringBitmap as a module via JPMS.

To accomplish this I added the `org.javamodularity.moduleplugin` plugin which allows the project to support Java 8 while still providing a module file. I also had to combine the classes from the `shims` project with the `RoaringBitmap` project and drop the `shims` project completely.

With this PR it is no longer possible to _compile_ the project without JDK 11. However, the project can still be **run** with Java 8 and it can also be tested with Java 8 using toolchains via `-PtestOnJava=8`. I also consolidated the Java pipelines workflows to test all versions at once.

### Automated Checks

- [👍] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [👍] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -61,6 +64,7 @@ subprojects.filter { !listOf("jmh", "fuzz-tests", "examples", "bsi", "simplebenc
                 configFile = File(rootProject.projectDir, "RoaringBitmap/style/roaring_google_checks.xml")
                 isIgnoreFailures = false
                 isShowViolations = true
+                exclude("module-info.java")
```
Comment (stiemannkj1) [inline build.gradle.kts:67]: `module-info.java` breaks checkstyle for some reason.

File: RoaringBitmap/build.gradle.kts [hunk 2]
```diff
@@ -14,6 +53,32 @@ dependencies {
...
+        from(sourceSets.named("java11").get().output)
+    }
+    manifest.attributes(
+            Pair("Multi-Release", "true")
+    )
+
+    // normally jar is just main classes but we also have another sourceset
+    dependsOn(tasks.named("compileJava11Java"))
+}
+
```
Comment (stiemannkj1) [inline RoaringBitmap/build.gradle.kts:81]: This code was moved from the `shims` build file.

File: .github/workflows/java-all-versions.yml
Comment (stiemannkj1) [inline .github/workflows/java-all-versions.yml:1]: Should probably modify the README to mention that JDK 11 is required to compile the project.

File: RoaringBitmap/src/main/java/org/roaringbitmap/insights/NaiveWriterRecommender.java [hunk 3]
```diff
@@ -78,7 +78,7 @@ private static void containerCountRecommendations(BitmapStatistics basedOn, Stri
       .append(averageContainersCount)
       .append("), because on average each bitmap has ")
       .append(averageContainersCount)
-      .append(" containers.\n");
+        .append(" containers.\n");
```
Comment (stiemannkj1) [inline RoaringBitmap/src/main/java/org/roaringbitmap/insights/NaiveWriterRecommender.java:81]: Not sure why, but this file failed `checkstyleMain` after my changes. So I fixed the indentation as checkstyle required.

File: RoaringBitmap/build.gradle.kts [hunk 4]
```diff
@@ -1,8 +1,47 @@
...
+tasks.test {
+    extensions.configure(org.javamodularity.moduleplugin.extensions.TestModuleOptions::class) {
+        setRunOnClasspath(true)
+    }
+}
+
+tasks.compileTestJava {
+    extensions.configure(org.javamodularity.moduleplugin.extensions.CompileTestModuleOptions::class) {
+        setCompileOnClasspath(true)
+    }
```
Comment (stiemannkj1) [inline RoaringBitmap/build.gradle.kts:39]: Add comments explaining that this avoids using modules for testing which allows the tests to run on Java 8.

File: RoaringBitmap/build.gradle.kts [hunk 5]
```diff
@@ -1,8 +1,47 @@
...
+    maven {
+      url = uri("https://plugins.gradle.org/m2/")
+    }
+  }
+  dependencies {
+    classpath("org.javamodularity:moduleplugin:1.8.12")
+  }
+}
+
+apply(plugin = "org.javamodularity.moduleplugin")
```
Comment (stiemannkj1) [inline RoaringBitmap/build.gradle.kts:12]: I'll add comments explaining this plugin.

**Comment by lemire** [comment]
Note that JDK 11+ is required to build the project already.

**Comment by lemire** [comment]
I am getting that module-info.class (and not just module-info.java) ends up in the source Jar files. Is this expected?

**Review by lemire — APPROVED** [review]
Looks good to me.

Can you just tell me whether it is expected that module-info.class should end up in the sources Jar? That seems awkward.

**Comment by stiemannkj1** [comment]
Thanks!

> Can you just tell me whether it is expected that module-info.class should end up in the sources Jar? That seems awkward.

Yeah that seems wrong 😅 I'll fix that.

**Comment by stiemannkj1** [comment]
@lemire had to hack around the issue as it's a bug in the plugin: https://github.com/java9-modularity/gradle-modules-plugin/issues/220.

See: https://github.com/RoaringBitmap/RoaringBitmap/pull/657/files#diff-befe560374c0f8a31ca75af604079f8c5836a487b472cf874ebc6e356395f379R16

Everything should be good now.

**Comment by lemire** [comment]
Ok. Looks good. Merging. I will issue a release.

**Pull request merged**